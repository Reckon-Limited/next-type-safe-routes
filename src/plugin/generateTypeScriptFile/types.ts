export type Page = {
  route: string;
  params?: string[];
};

export type ApiRoute = {
  route: string;
  params?: string[];
};
