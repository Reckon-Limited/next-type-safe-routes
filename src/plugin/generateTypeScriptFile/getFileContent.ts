import { ApiRoute, Page } from "./types";

const getTypeSafeRoute = ({ route, params }: ApiRoute) => {
  if (!params?.length) {
    return `"${route}"`;
  }

  return `{ route: "${route}", ${params
    .map((param) => `${param}: string | string[] | number`)
    .join(",")} }`;
};

type Args = {
  apiRoutes: ApiRoute[];
  pages: Page[];
};

const getFileContent = ({
  apiRoutes,
  pages,
}: Args) => `/* eslint no-use-before-define: 0 */ // --> OFF
// prettier-ignore

// IMPORTANT! This file is autogenerated by the \`type-safe-next-routes\` 
// package. You should _not_ update these types manually...

declare module "next-type-safe-routes" {
  export type TypeSafePage = ${pages.map(getTypeSafeRoute).join(" | ")};
  ${
    apiRoutes.length > 0
      ? `export type TypeSafeApiRoute = ${apiRoutes
          .map(getTypeSafeRoute)
          .join(" | ")};`
      : ""
  }

  export const getPathname = (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
  export const getRoute = (typeSafeUrl: TypeSafePage | TypeSafeApiRoute, query?: any) => string;
}
`;

export default getFileContent;
