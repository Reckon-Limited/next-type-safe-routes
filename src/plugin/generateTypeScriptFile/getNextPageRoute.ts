const getNextPageRoute = (fileName: string) => {
  const route = fileName
    // remove the file extension
    .split(".")[0]
    // index pages don't need the "/index" when used as hrefs
    .replace("/index", "");

  // if this is the root index file, return "/"" instead of ""
  return route === "" ? "/" : route;
};

export default getNextPageRoute;
