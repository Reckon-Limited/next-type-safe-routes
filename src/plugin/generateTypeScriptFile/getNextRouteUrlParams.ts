const getNextRouteUrlParams = (href: string) => {
  const params = href.match(/\[([^\]]+)\]/g);
  return params?.map((param) => param.replace("[", "").replace("]", ""));
};

export default getNextRouteUrlParams;
