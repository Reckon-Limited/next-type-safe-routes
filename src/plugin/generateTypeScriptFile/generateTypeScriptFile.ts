import walkSync from "walk-sync";

import getFileContent from "./getFileContent";
import getRoutes from "./getRoutes";

const ignorePagesRoutes = ["/_app.tsx", "/_document.tsx"];
const shouldIncludePageEntry = (route: string) =>
  route.match(".tsx") && !ignorePagesRoutes.includes(route);
const shouldIncludeApiRouteEntry = (endpoint: string) => endpoint.match(".ts");

const generateTypeScriptFile = (pagesDir: string) => {
  const pagesFiles = walkSync(pagesDir, {
    directories: false,
    ignore: ["api"],
  });
  const apiRouteFiles = walkSync(`${pagesDir}/api`, {
    directories: false,
  });

  const relevantPages = pagesFiles.filter(shouldIncludePageEntry);
  const pages = getRoutes(relevantPages.map((page) => `/${page}`));
  const relavantApiRoutes = apiRouteFiles.filter(shouldIncludeApiRouteEntry);
  const apiRoutes = getRoutes(
    relavantApiRoutes.map((route) => `/api/${route}`)
  );

  const fileContent = getFileContent({ pages, apiRoutes });

  return fileContent;
};

export default generateTypeScriptFile;
