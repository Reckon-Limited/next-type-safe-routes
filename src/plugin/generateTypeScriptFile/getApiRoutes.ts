import getNextPageRoute from "./getNextPageRoute";
import getNextRouteUrlParams from "./getNextRouteUrlParams";
import { ApiRoute } from "./types";

const shouldIncludeEntry = (endpoint: string) => endpoint.match(".ts");

const getApiRoutes = (fileNames: string[]): ApiRoute[] => {
  console.log({ fileNames });
  return fileNames.filter(shouldIncludeEntry).map((fileName) => {
    const route = getNextPageRoute(fileName);
    const params = getNextRouteUrlParams(route);
    return { route, params };
  });
};

export default getApiRoutes;
