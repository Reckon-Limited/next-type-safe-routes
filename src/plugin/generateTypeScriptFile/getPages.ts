import getNextPageRoute from "./getNextPageRoute";
import getNextRouteUrlParams from "./getNextRouteUrlParams";
import { Page } from "./types";

const ignoreRoutes = ["/_app.tsx", "/_document.tsx"];

const shouldIncludeEntry = (route: string) =>
  route.match(".tsx") && !ignoreRoutes.includes(route);

const getPages = (fileNames: string[]): Page[] => {
  return fileNames.filter(shouldIncludeEntry).map((fileName) => {
    const route = getNextPageRoute(fileName);
    const params = getNextRouteUrlParams(route);
    return { route, params };
  });
};

export default getPages;
