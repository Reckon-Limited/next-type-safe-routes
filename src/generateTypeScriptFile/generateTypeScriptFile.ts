import walkSync from "walk-sync";

import getApiRoutes from "./getApiRoutes";
import getFileContent from "./getFileContent";
import getPages from "./getPages";

const generateTypeScriptFile = (pagesDir: string) => {
  const pagesFiles = walkSync(pagesDir, {
    directories: false,
    ignore: ["api"],
  });
  const apiRouteFiles = walkSync(`${pagesDir}/api`, { directories: false });

  const pages = getPages(pagesFiles);
  const apiRoutes = getApiRoutes(apiRouteFiles);
  const fileContent = getFileContent({ pages, apiRoutes });

  return fileContent;
};

export default generateTypeScriptFile;
