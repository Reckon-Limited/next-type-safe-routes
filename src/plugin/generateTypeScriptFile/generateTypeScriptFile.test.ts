import generateTypeScriptFile from "./index";

const pagesDir = __dirname + "/mocks/pages";

describe("plugin/generateTypeScriptFile", () => {
  it("works as expected", () => {
    expect(generateTypeScriptFile(pagesDir)).toMatchSnapshot();
  });
});
