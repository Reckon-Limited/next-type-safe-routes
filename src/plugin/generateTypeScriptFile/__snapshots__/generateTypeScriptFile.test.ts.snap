// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin/generateTypeScriptFile works as expected 1`] = `
"/* eslint no-use-before-define: 0 */ // --> OFF
// prettier-ignore

// IMPORTANT! This file is autogenerated by the \`type-safe-next-pages\` 
// package. You should _not_ update these types manually...

declare module \\"next-type-safe-routes\\" {
  export type TypeSafePage = \\"/404\\" | \\"/\\" | { route: \\"/users/[userId]\\", userId: string | string[] | number } | \\"/users\\";
  export type TypeSafeApiRoute = { route: \\"/api/[authId]\\", authId: string | string[] | number } | { route: \\"/api/users/[userId]\\", userId: string | string[] | number } | \\"/api/users\\";

  export const getPathname = (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
  export const getRoute = (typeSafeUrl: TypeSafePage | TypeSafeApiRoute, query?: any) => string;
}
"
`;
